### DISCUSSION TITLE
How to Read a .mf4 measurement file and access its contents

### DISCUSSION THREAD

**Discussion by sterfrag** [ihedvall/mdflib#74] [Show and tell]
Hello and thank you very much for your work.
I have downloaded your project and I am trying to read an .mf4 file with measurements from our hil. Is there a way, or approach that you use in order to access the measurements and signals that have been recorded?

I enclose the code I use in order to open the mf4 file, read its datagroups and channelgroups and so on

```
int main() {
    // Specify the path to the MDF file
    std::string mdfFilePath {"C:/Users/s.fragkos/Desktop/Recording.mf4"};
    FILE*  mdfile = fopen(mdfFilePath.c_str(),"rb");

    // Create an instance of MdfReader to open the MDF file
    mdf::MdfReader mdfReader(mdfFilePath);
    mdfReader.Open();
    const mdf::MdfFile* ptr_MdfFile = mdfReader.GetFile();
        
    mdfReader.ReadMeasurementInfo();
    
    const mdf::IHeader* ptr_header = mdfReader.GetHeader();
    
    std::string mdf_version = ptr_MdfFile->Version();    
    
    mdf::DataGroupList dest2;

    ptr_MdfFile->DataGroups(dest2);
    std::cout<<mdfReader.ReadData(*(dest2.at(0)))<<std::endl;
    mdfReader.ReadData(*(dest2.at(0)));

    if ( mdfReader.ReadData(*(dest2.at(0))))
    {
        std::cout<< "DataGroup was read succesfully"<<std::endl;
        std::cout<<"DGList contains "<<dest2.size()<<" DataGroup"<<std::endl;
        mdf::IDataGroup* SingleDataGroup = dest2.at(0);

        std::cout<<"1 DataGroup contains "<<SingleDataGroup->ChannelGroups().size()<<" ChannelGroups"<<std::endl;

        mdf::IChannelGroup* SingleChannelGroup = SingleDataGroup->GetChannelGroup(0);
        std::cout<<"Number of Channels inside ChannelGroup: "<<SingleChannelGroup->Channels().size()<<std::endl;

        std::cout<<SingleChannelGroup->NofSamples()<< " is the number of samples found in this ChannelGroup"<<std::endl;
        mdf::SampleRecord SampleRecordFromChannelGroup = SingleChannelGroup->GetSampleRecord();
        mdf::ISourceInformation* SourceInfo = SingleChannelGroup->SourceInformation();

        bool bfr_vector = SampleRecordFromChannelGroup.vlsd_data;

        std::cout << bfr_vector<<std::endl;
    }
```
.
When this code is run, the ouput is:
```
1
DataGroup was read succesfully
DGList contains 1 DataGroup
1 DataGroup contains 1 ChannelGroups
Number of Channels inside ChannelGroup: 0
14500 is the number of samples found in this ChannelGroup
0
Started processing...
MDF File has datagroups and procedure was correct:

``` 
. Is there a more appropriate approach in how to access the measurements?

**Comment by ihedvall** [comment]
Yes it is simpler and more complicated ways to read an MDF4 file. But first, there is an online help about the MDF library at [https://ihedvall.github.io/mdflib/]( https://ihedvall.github.io/mdflib/). There is an simple MDF Reader example. You need to do the following steps.

1. Create an MdfReader object including the path to the file. Note that the file is now open.
2. Call the function ReadEverythingButData(). This reads in all file information about the contents but doesn't read in the data values. 
3. The next step is to select which channel data to read. Note that reading all data may consume memory, so you should only read channel data needed. You need to create so called channel observers (see the help example). The observer holds the values for each sample and handle the conversion to engineering values.
4. Now you should call the ReadData() function. You need to supply which data group i.e. measurement, you want to read. This populates the observers with sample data.
5. The observers now have all data of interest. It is up to you what to do with the data.

There exists more complicated ways to fetch data but let's start with a small baby steps. Note that it is the ReadData() function that consumes time and memory.

**Comment by sterfrag** [comment]
Thank you very much for your help. Although it helped me, I still cannot read the values inside the mf4 file. My new code is 

```
int main() {
    // Specify the path to the MDF file
    std::string mdfFilePath {"C:/Users/s.fragkos/Desktop/mpu_resets.mf4"};

    // Create an instance of MdfReader to open the MDF file
    mdf::MdfReader mdfReader(mdfFilePath);
    mdfReader.ReadEverythingButData();
    
    //mdfReader.Open();
    
    const mdf::MdfFile* ptr_MdfFile = mdfReader.GetFile(); 
    
    std::string mdf_version = ptr_MdfFile->Version();    
    
    mdf::DataGroupList dest2;

    ptr_MdfFile->DataGroups(dest2);
    mdf::IDataGroup* SingleDataGroup = dest2.at(1);
    
    if ( mdfReader.ReadData(*SingleDataGroup))
    {
        mdfReader.ReadData(*(SingleDataGroup));
        std::cout<< "DataGroup was read succesfully"<<std::endl;
        std::cout<<"DGList contains "<<dest2.size()<<" DataGroup/s"<<std::endl;
        

        std::cout<<"1 DataGroup contains "<<SingleDataGroup->ChannelGroups().size()<<" ChannelGroups"<<std::endl;


        mdf::IChannelGroup* SingleChannelGroup = SingleDataGroup->GetChannelGroup(0);

        std::vector<mdf::IChannel*> ChannelList = SingleChannelGroup->Channels();
        mdf::IChannel* SingleChannel = ChannelList.at(11);
        //
        //mdf::IChannelArray* CHArrayobj = SingleChannel->ChannelArray();

        //std::cout<<"Dimensions of ChannelArray object are: "<<CHArrayobj->Dimensions()<<std::endl;

        std::cout<<SingleChannelGroup->NofSamples()<< " is the number of samples found in this ChannelGroup"<<std::endl;
        mdf::ChannelObserverPtr sub = mdf::CreateChannelObserver(*(SingleDataGroup),*SingleChannelGroup,*(SingleChannel));
        sub->ReadVlsdData(true);
        std::string channel_value ; // Channel value (no scaling)
        double eng_value ; // Engineering value
        for (size_t sample=0; sample < (sub->NofSamples()) ; ++sample)
        {   
            SingleChannel->GetChannelValue()
            const auto channel_valid = sub->GetChannelValue(sample, channel_value);
            const auto eng_valid = sub->GetEngValue(sample,eng_value);
            std::string eng_Val_to_string = sub->EngValueToString(sample);
            std::cout<<"Value in sample: "<<sample<< " is "<< eng_valid<< ","<< channel_valid <<std::endl;
            std::cout<<"Value in sample: "<<sample<< " converted to string is: "<< eng_Val_to_string <<std::endl;
        }
        std::cout<<"Number of Channels inside ChannelGroup: "<<SingleChannelGroup->Channels().size()<<std::endl;
        for (size_t numofchannel=0;numofchannel<ChannelList.size()-1;++numofchannel)
        {
            std::cout<<"Name of "<<numofchannel <<" Channel is: "<<ChannelList.at(numofchannel)->Name()<<std::endl;
        }
    }

```

and the output is something like.

```
Value in sample: 2199 converted to string is: *
Value in sample: 2200 is 0,0
Value in sample: 2200 converted to string is: *
Number of Channels inside ChannelGroup: 39
Name of 0 Channel is: time
Name of 1 Channel is: Icsp_Dem_Fmy_EventId[7]
```

**Comment by ihedvall** [comment]
The main issue is that you haven't added any channel subscribers before calling the ReadData() function. That function read sample by sample and calls the channel subscriber(s) that in turn save the sample for later usage.

The ReadData() function handles the complex with VLSD data. There is no meaning to call any more ReadXXX() function unless the data doesn't fit the memory (large files typical with video streams).

I return soon with a rewritten code.

**Comment by sterfrag** [comment]
Aaah I see, yes now it is solved! Thank you very much! I read the values appropriately now

**Comment by ihedvall** [comment]
I have rewritten your code below. I moved half of the code inside the if statement, to before the ReadData() call. Note that the subscriber is
added before the ReadData() call. The subscriber will have all sample values.

The above code may contains some minor compiler errors. I have not tested it. 
```
#include <string_view>
#include <string>
#include <vector>
#include <iostream>

int main() {
    // Specify the path to the MDF file
    const std::string_view mdfFilePath ="C:/Users/s.fragkos/Desktop/mpu_resets.mf4";

    // Create an instance of MdfReader to open the MDF file
    mdf::MdfReader mdfReader(mdfFilePath);
    mdfReader.ReadEverythingButData();
    
    const auto* ptr_MdfFile = mdfReader.GetFile(); 
    const auto mdf_version = ptr_MdfFile->Version();    
    
    mdf::DataGroupList dest2;

    ptr_MdfFile->DataGroups(dest2);
    std::cout<<"DGList contains "<<dest2.size()<<" DataGroup/s"<<std::endl;

    // Note: This is the second data block (DG)
    mdf::IDataGroup* SingleDataGroup = dest2.at(1);
    std::cout <<"1 DataGroup contains "<<SingleDataGroup->ChannelGroups().size()<<" ChannelGroups"<<std::endl;

    mdf::IChannelGroup* SingleChannelGroup = SingleDataGroup->GetChannelGroup(0);
    std::cout << SingleChannelGroup->NofSamples()
              << " is the number of samples found in this ChannelGroup"<<std::endl;
    std::cout<<"Number of Channels inside ChannelGroup: "
              << SingleChannelGroup->Channels().size() <<std::endl;
    
    std::vector<mdf::IChannel*> ChannelList = SingleChannelGroup->Channels();

    mdf::IChannel* SingleChannel = ChannelList.at(11);

    mdf::ChannelObserverPtr sub = mdf::CreateChannelObserver(*SingleDataGroup,
                                                             *SingleChannelGroup,
                                                             *SingleChannel);

    mdfReader.ReadData(*SingleDataGroup))
    std::cout<< "DataGroup was read succesfully"<<std::endl;


    std::string channel_value ; // Channel value (no scaling)
    double eng_value ; // Engineering value
    for (size_t sample=0; sample < (sub->NofSamples()) ; ++sample)
    {   
      // Note that it is the subscriber that holds the values.
      const auto channel_valid = sub->GetChannelValue(sample, channel_value);
      const auto eng_valid = sub->GetEngValue(sample,eng_value);
      
      // Note: Better to call GetEngValue with a string. It will convert it 
      // to a string.
      std::string eng_Val_to_string = sub->EngValueToString(sample);
      std::cout<<"Value in sample: "<< sample<< " is "<< eng_valid<< "," << channel_valid << std::endl;
      std::cout<<"Value in sample: "<< sample<< " converted to string is: " << eng_Val_to_string <<std::endl;
    }
}
```

**Reply by sterfrag** [reply]
Thanks a lot, this was very helpful and I am able now to read successfully the values of channels. 
Is there also a way to map these values that are read, to a specific measurement time? From my channels, I noticed that the 1st channel is always the time of the measurement. I want this, because some of my measurements have different sampling time.

**Comment by ihedvall** [comment]
You are actually spot-on with the weakness of the MDF specification. Each CG group consist one or more channels. One of these channels (normally the first one) is marked as a master channel. The master channel is normally relative time but note that it may also be relative to degree and distance. The master channel is typical plotted on the X-axis.

The start time is official stored in the HD (header block) and is nano-seconds since 1970. Unfortunately, this implies that the file only have one measurement. Many "viewers" assumes this. 

There exist some other solution as storing the start/stop time as meta data (MD block) or as events (EV block). None of them are standard.

**Reply by sterfrag** [reply]
Thank you very much for your response. This was most helpful

**Comment by bekiraydin** [comment]
> Hello and thank you very much for your work. I have downloaded your project and I am trying to read an .mf4 file with measurements from our hil. Is there a way, or approach that you use in order to access the measurements and signals that have been recorded?
> 
> I enclose the code I use in order to open the mf4 file, read its datagroups and channelgroups and so on
> 
> ```
> int main() {
>     // Specify the path to the MDF file
>     std::string mdfFilePath {"C:/Users/s.fragkos/Desktop/Recording.mf4"};
>     FILE*  mdfile = fopen(mdfFilePath.c_str(),"rb");
> 
>     // Create an instance of MdfReader to open the MDF file
>     mdf::MdfReader mdfReader(mdfFilePath);
>     mdfReader.Open();
>     const mdf::MdfFile* ptr_MdfFile = mdfReader.GetFile();
>         
>     mdfReader.ReadMeasurementInfo();
>     
>     const mdf::IHeader* ptr_header = mdfReader.GetHeader();
>     
>     std::string mdf_version = ptr_MdfFile->Version();    
>     
>     mdf::DataGroupList dest2;
> 
>     ptr_MdfFile->DataGroups(dest2);
>     std::cout<<mdfReader.ReadData(*(dest2.at(0)))<<std::endl;
>     mdfReader.ReadData(*(dest2.at(0)));
> 
>     if ( mdfReader.ReadData(*(dest2.at(0))))
>     {
>         std::cout<< "DataGroup was read succesfully"<<std::endl;
>         std::cout<<"DGList contains "<<dest2.size()<<" DataGroup"<<std::endl;
>         mdf::IDataGroup* SingleDataGroup = dest2.at(0);
> 
>         std::cout<<"1 DataGroup contains "<<SingleDataGroup->ChannelGroups().size()<<" ChannelGroups"<<std::endl;
> 
>         mdf::IChannelGroup* SingleChannelGroup = SingleDataGroup->GetChannelGroup(0);
>         std::cout<<"Number of Channels inside ChannelGroup: "<<SingleChannelGroup->Channels().size()<<std::endl;
> 
>         std::cout<<SingleChannelGroup->NofSamples()<< " is the number of samples found in this ChannelGroup"<<std::endl;
>         mdf::SampleRecord SampleRecordFromChannelGroup = SingleChannelGroup->GetSampleRecord();
>         mdf::ISourceInformation* SourceInfo = SingleChannelGroup->SourceInformation();
> 
>         bool bfr_vector = SampleRecordFromChannelGroup.vlsd_data;
> 
>         std::cout << bfr_vector<<std::endl;
>     }
> ```
> 
> . When this code is run, the ouput is:
> 
> ```
> 1
> DataGroup was read succesfully
> DGList contains 1 DataGroup
> 1 DataGroup contains 1 ChannelGroups
> Number of Channels inside ChannelGroup: 0
> 14500 is the number of samples found in this ChannelGroup
> 0
> Started processing...
> MDF File has datagroups and procedure was correct:
> ```
> 
> . Is there a more appropriate approach in how to access the measurements?

Hello , I create a Python GUI for data reading and plotting. If you are interested I can send you. I only need to hear some feedbacks from real users :)

**Reply by ihedvall** [reply]
@bekiraydin 
Your code have some error. 
A simple example of an MDF read is available in the documentation [MDF File Read](https://ihedvall.github.io/mdflib/mdfreader.html).
Reading is done in 4 basic steps.

1. Initialize a MDF Reader object and define the path to the file.
2. Call the ReadEverythingButData() function. Now, you know all configuration.
3. Create Channel Observers for each channel that is wanted. 
4. Call the ReadData() function for each data group in the file. Repeat [3] and [4] for each data group.

It is the 3:rd step that is missing in your code. 
The purpose of a channel observer is to store all samples for a channel in the primary memory.
If you want to plot data, it is the channel observers that you should ask.
There are some other errors like calling ReadMeasurementInfo() instead of [2] and calling ReadData() twice for a data group.

If you want a fast look of the sample data you can use the MDF Viewer application in this repository [Release Area](https://github.com/ihedvall/mdflib/releases/tag/v2.3.0).
You can also use the [ASAMMDF Python GUI/Library](https://github.com/danielhrisca/asammdf).

An MDF file contains measurement (sample) data but also information about the test object.
Reading the sample data take longer time that to read the configuration information. 
If you have >1GB files, there exist some other functions as your sample data doesn't fit into the memory but let us focus on your simple example.

Please return if you have some more questions or you want some feedback of your application.
